pkg/kubectl/cmd/apply/prune/prune.go

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prune
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+)
+
+type Pruner struct {
+	Mapper               meta.RESTMapper
+	GroupVersionKinds    []*meta.RESTMapping
+	ClientFunc           resource.ClientMapperFunc
+	ClientsetFunc        func() (internalclientset.Interface, error)
+	Selector             string
+	IncludeUninitialized bool
+	Cascade              bool
+	GracePeriod          int
+	DryRun               bool
+	Out                  io.Writer
+	ShortOutput          bool
+}
+
+// Param: gvks array of groupVersionKind passed in --prune-whitelist or empty array if no flag.
+// Returns default groupVersionKind mappings to prune, unless overriden by --prune-whitelist flag.
+func ParseGvks(mapper meta.RESTMapper, gvks []string) ([]*meta.RESTMapping, error) {
+
+	// QUESTION: Do I need to explicitly define if namespaced?
+	// Is correct namespace available from "mapping.Scope.Name()"?
+	// Default group/version/kind to prune
+	pruneGvks := [][]string{
+		{"", "v1", "ConfigMap"},
+		{"", "v1", "Endpoints"},
+		{"", "v1", "Namespace"},
+		{"", "v1", "PersistentVolumeClaim"},
+		{"", "v1", "PersistentVolume"},
+		{"", "v1", "Pod"},
+		{"", "v1", "ReplicationController"},
+		{"", "v1", "Secret"},
+		{"", "v1", "Service"},
+		{"batch", "v1", "Job"},
+		{"extensions", "v1beta1", "DaemonSet"},
+		{"extensions", "v1beta1", "Deployment"},
+		{"extensions", "v1beta1", "Ingress"},
+		{"extensions", "v1beta1", "ReplicaSet"},
+		{"apps", "v1beta1", "StatefulSet"},
```

Comment:
These are all out of date.  This strategy was always a bad idea...

Reply:
Let's keep the exact same code, and not change the functionality in this PR, in order to make the PR low risk. Once merged I'll open an issue and address it in a future PR.